components/ai_chat/core/browser/associated_content_manager.cc

```diff
@@ -0,0 +1,399 @@
+          content_callback));
+    }
+  } else {
+    on_page_text_fetch_complete_->Post(FROM_HERE, std::move(callback));
+  }
+}
+
+void AssociatedContentManager::GetScreenshots(
+    ConversationHandler::GetScreenshotsCallback callback) {
+  DVLOG(1) << __func__;
+
+  if (content_delegates_.size() == 0) {
+    std::move(callback).Run(std::nullopt);
+    return;
+  }
+
+  auto all_screenshots =
+      base::BarrierCallback<std::optional<std::vector<mojom::UploadedFilePtr>>>(
+          content_delegates_.size(),
+          base::BindOnce(
+              [](ConversationHandler::GetScreenshotsCallback callback,
+                 std::vector<std::optional<std::vector<mojom::UploadedFilePtr>>>
+                     screenshots) {
+                bool all_nullopt = true;
+                std::vector<mojom::UploadedFilePtr> all_screenshots;
+                for (auto& set : screenshots) {
+                  if (!set) {
+                    continue;
+                  }
+
+                  all_nullopt = false;
+                  std::ranges::move(set.value(),
+                                    std::back_inserter(all_screenshots));
+                }
+                if (all_nullopt) {
+                  std::move(callback).Run(std::nullopt);
+                } else {
+                  std::move(callback).Run(std::move(all_screenshots));
+                }
+              },
+              std::move(callback)));
+
+  for (auto* content : content_delegates_) {
+    content->GetScreenshots(all_screenshots);
+  }
+}
+
+void AssociatedContentManager::GetStagedEntriesFromContent(
+    ConversationHandler::GetStagedEntriesCallback callback) {
+  DVLOG(1) << __func__;
+
+  if (content_delegates_.size() != 1) {
+    std::move(callback).Run(std::nullopt);
+    return;
+  }
+
+  content_delegates_[0]->GetStagedEntriesFromContent(std::move(callback));
+}
+
+void AssociatedContentManager::GetTopSimilarityWithPromptTilContextLimit(
+    const std::string& prompt,
+    const std::string& text,
+    uint32_t context_limit,
+    TextEmbedder::TopSimilarityCallback callback) {
+  // TODO(https://github.com/brave/brave-browser/issues/45731): Remove this
+  // function
+  DVLOG(1) << __func__;
+
+  if (content_delegates_.size() != 1) {
+    std::move(callback).Run(base::unexpected("No content delegate"));
+    return;
+  }
+
+  content_delegates_[0]->GetTopSimilarityWithPromptTilContextLimit(
+      prompt, text, context_limit, std::move(callback));
+}
+
+std::string AssociatedContentManager::GetCachedTextContent() const {
+  DVLOG(1) << __func__;
+
+  auto cached_content = GetCachedContent();
+
+  // If we only have one content delegate directly return the content.
+  // Otherwise, wrap each content in <page> tags.
+  if (cached_content.size() == 1) {
+    return std::string(cached_content[0]);
+  }
+
+  if (cached_content.size() > 0) {
+    return base::StrCat({"<page>",
+                         base::JoinString(cached_content, "</page><page>"),
```

Comment:
Shouldn't we strip this delimiter from each content?

Reply:
https://github.com/brave/brave-core/pull/28483/commits/5f1a5386132ff3623c452e4a4cc766a7416f6872